cmd/cloud-controller-manager/app/options/options.go

```diff
@@ -226,11 +248,138 @@ func (o *CloudControllerManagerOptions) ApplyTo(
 	// sync back to component config
 	// TODO: find more elegant way than syncing back the values.
 
+	setUpBlockAffinitiesInformer(c.DynamicInformers, c.SharedInformers)
+
 	c.ComponentConfig.NodeStatusUpdateFrequency = o.NodeStatusUpdateFrequency
 
 	return nil
 }
 
+type NodePodCIDR struct {
+	NodeIP  string `json:"nodeIP"`
+	PodCIDR string `json:"podCIDR"`
+	Action  string `json:"action"`
+}
+
+func getNodePrivateIPAddresses(node *v1.Node) []string {
+	addresses := make([]string, 0)
+	for _, nodeAddress := range node.Status.Addresses {
+		if strings.EqualFold(string(nodeAddress.Type), string(v1.NodeInternalIP)) {
+			addresses = append(addresses, nodeAddress.Address)
+		}
+	}
+
+	return addresses
+}
+
+func setUpBlockAffinitiesInformer(dynamicInformerFactory dynamicinformer.DynamicSharedInformerFactory, informerFactory informers.SharedInformerFactory) {
+
+	klog.Info("Setting up Block Affinities Informer")
+
+	resource := schema.GroupVersionResource{Group: "crd.projectcalico.org", Version: "v1", Resource: "blockaffinities"}
+	informer := dynamicInformerFactory.ForResource(resource).Informer()
+
+	nodeLister := informerFactory.Core().V1().Nodes().Lister()
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			typedObj := obj.(*unstructured.Unstructured)
+
+			bytes, _ := typedObj.MarshalJSON()
+
+			var blockAffinity *v3.BlockAffinity
+			json.Unmarshal(bytes, &blockAffinity)
+
+			podCIDR := blockAffinity.Spec.CIDR
+			nodeName := blockAffinity.Spec.Node
+
+			node, err := nodeLister.Get(nodeName)
+
+			if err != nil {
+				klog.Warning(err)
+			} else {
+				putUdr(node, podCIDR, "add")
+			}
+
+			klog.Infof("BlockAffinity added: %v", blockAffinity)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+
+			typedObj := newObj.(*unstructured.Unstructured)
+			bytes, _ := typedObj.MarshalJSON()
+
+			var blockAffinity *v3.BlockAffinity
+			json.Unmarshal(bytes, &blockAffinity)
+
+			podCIDR := blockAffinity.Spec.CIDR
+			nodeName := blockAffinity.Spec.Node
+
+			node, err := nodeLister.Get(nodeName)
+
+			if err != nil {
+				klog.Warning(err)
+			} else {
+				putUdr(node, podCIDR, "update")
+			}
+
+			klog.Infof("BlockAffinity updated: %v", blockAffinity.Spec.CIDR)
+		},
+		DeleteFunc: func(obj interface{}) {
+
+			typedObj := obj.(*unstructured.Unstructured)
+			bytes, _ := typedObj.MarshalJSON()
+
+			var blockAffinity *v3.BlockAffinity
+			json.Unmarshal(bytes, &blockAffinity)
+
+			podCIDR := blockAffinity.Spec.CIDR
+			nodeName := blockAffinity.Spec.Node
+
+			node, err := nodeLister.Get(nodeName)
+
+			if err != nil {
+				klog.Warning(err)
+			} else {
+				putUdr(node, podCIDR, "delete")
+			}
+
+			klog.Infof("BlockAffinity deleted: %v", blockAffinity)
+		},
+	})
+
+}
+
```

Comment:
Generally, good to put atleast a one-line comment describing the function. You could ignore this comment as it's not product code.